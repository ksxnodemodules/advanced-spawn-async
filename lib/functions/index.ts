import core from './core'
import main from './main'
export { core, main }
