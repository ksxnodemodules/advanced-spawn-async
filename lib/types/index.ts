import { SpawnOptions } from 'child_process'

export interface IsomorphicSpawn<Return extends IsomorphicSpawn.Return> {
  (command: string, args: string[], options: SpawnOptions): Return
}

export namespace IsomorphicSpawn {
  export interface Return {
    on (event: 'close' | 'exit', listener: (status: number, signal: string | null) => void): void
    on (event: 'error', listener: (error: Error) => void): void
    stdout: Readable
    stderr: Readable
  }

  export interface Readable {
    on (event: 'data', listener: (chunk: Buffer) => void): void
  }
}

export interface SpawnFactory<Process extends IsomorphicSpawn.Return> {
  readonly onexit: SpawnFactory.TerminationPromise<Process>
  readonly onclose: SpawnFactory.TerminationPromise<Process>
  readonly process: Process
}

export namespace SpawnFactory {
  export type TerminationPromise<Process extends IsomorphicSpawn.Return> =
    Promise<TerminationInformation<Process>>

  export interface TerminationInformation<Process extends IsomorphicSpawn.Return> {
    readonly command: string
    readonly args: string[]
    readonly options: SpawnOptions
    readonly stdout: string
    readonly stderr: string
    readonly output: string
    readonly status: number
    readonly signal: string | null
    readonly process: Process
  }
}

export interface Options extends SpawnOptions {
  readonly event?: Options.TerminationEvent
}

export namespace Options {
  export type TerminationEvent = 'close' | 'exit'
}

export interface InternalErrorInformation<Process extends IsomorphicSpawn.Return, Error> {
  readonly command: string
  readonly args: string[]
  readonly options: SpawnOptions
  readonly process: Process
  readonly error: Error
}
