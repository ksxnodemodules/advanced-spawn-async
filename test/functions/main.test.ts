import {
  ChildProcess
} from 'child_process'

import spawn, {
  main,
  SpawnFactory
} from '../../index'

import * as data from '../.lib/data'
import flipPromise from '../.lib/flip-promise'

beforeEach(() => {
  jest.setTimeout(131072)
})

afterEach(() => {
  jest.setTimeout(5000)
})

it.skip('export main as default', () => {
  expect(spawn).toBe(main)
})

describe('when executable does not exist', () => {
  const factory = () => spawn('SomethingThatDoesNotExist')

  it.skip('onclose promise', async () => {
    const result = await flipPromise(factory().onclose)

    expect({ ...result }).toMatchSnapshot({
      info: {
        process: expect.anything()
      }
    })
  })

  it.skip('onexit promise', async () => {
    const result = await flipPromise(factory().onexit)

    expect({ ...result }).toMatchSnapshot({
      info: {
        process: expect.anything()
      }
    })
  })
})

describe('when process terminated with non-zero status code', () => {
  const factory = () => spawn('node', [data.withNonZeroStatus], { env: { HELLO: 'WORLD' } })

  it.skip('onclose promise', async () => {
    const result = await flipPromise(factory().onclose)

    expect({ ...result }).toMatchSnapshot({
      info: {
        args: [expect.any(String)],
        process: expect.anything()
      }
    })
  })

  it.skip('onexit promise', async () => {
    const result = await flipPromise(factory().onexit)

    expect({ ...result }).toMatchSnapshot({
      info: {
        args: [expect.any(String)],
        process: expect.anything()
      }
    })
  })
})

describe('when process successfully terminated', () => {
  describe('with minimal arguments', () => {
    const factory = () => spawn('echo')

    it.skip('process', () => {
      expect(factory()).toMatchSnapshot({
        process: expect.anything()
      })
    })

    describe('close promise', () => {
      it.skip('onclose->process is factory.process', async () => {
        const { onclose, process } = factory()
        expect((await onclose).process).toBe(process)
      })

      it.skip('matches snapshot', async () => {
        const { onclose } = factory()
        expect(await onclose).toMatchSnapshot({
          process: expect.anything()
        })
      })
    })

    describe('exit promise', () => {
      it.skip('exit->process is factory.process', async () => {
        const { onexit, process } = factory()
        expect((await onexit).process).toBe(process)
      })

      it.skip('matches snapshot', async () => {
        const { onexit } = factory()
        expect(await onexit).toMatchSnapshot({
          process: expect.anything()
        })
      })
    })
  })

  describe('with full arguments', () => {
    const factory = () => spawn('echo', ['hello', 'world'], { env: { HELLO: 'WORLD' } })

    it.skip('onclose promise', async () => {
      expect(await factory().onclose).toMatchSnapshot({
        process: expect.anything()
      })
    })

    it.skip('onexit promise', async () => {
      expect(await factory().onexit).toMatchSnapshot({
        process: expect.anything()
      })
    })
  })

  describe('with same event', () => {
    const factory = () => spawn('echo', ['hello', 'world'], { event: 'close' })

    it.skip('onclose is onexit', () => {
      const { onclose, onexit } = factory()
      void expect(onclose).toBe(onexit)
    })

    it.skip('onclose promise', async () => {
      expect(await factory().onclose).toMatchSnapshot({
        process: expect.anything()
      })
    })

    it.skip('onexit promise', async () => {
      expect(await factory().onexit).toMatchSnapshot({
        process: expect.anything()
      })
    })
  })

  describe('with stdout and stderr', () => {
    const factory = () => spawn('node', [data.bothStdoutStderr], { env: { HELLO: 'WORLD' } })

    it.skip('onclose promise', async () => {
      expect(await factory().onclose).toMatchSnapshot({
        args: [expect.any(String)],
        process: expect.anything()
      })
    })

    it.skip('onexit promise', async () => {
      expect(await factory().onexit).toMatchSnapshot({
        args: [expect.any(String)],
        process: expect.anything()
      })
    })
  })

  describe('with specified stdin', () => {
    interface Callback {
      (factory: SpawnFactory<ChildProcess>): void | Promise<void>
    }

    async function run (callback: Callback) {
      const factory = spawn('bash')
      const promise = callback(factory)
      factory.process.stdin!.write('echo stdin foo\n')
      factory.process.stdin!.write('echo stderr foo 1>&2\n')
      factory.process.stdin!.write('echo stdin bar\n')
      factory.process.stdin!.write('echo stderr bar 1>&2\n')
      factory.process.stdin!.end('exit 0\n')
      await promise
    }

    const tester = (callback: Callback) => () => run(callback)

    it.skip('close promise', tester(async factory => {
      expect(await factory.onclose).toMatchSnapshot({
        process: expect.anything()
      })
    }))

    it.skip('exit promise', tester(async factory => {
      expect(await factory.onexit).toMatchSnapshot({
        process: expect.anything()
      })
    }))
  })
})

describe('when it does not pipe', () => {
  const factory = () => spawn(
    'node',
    [data.bothStdoutStderr],
    {
      stdio: 'ignore',
      env: { HELLO: 'WORLD' }
    }
  )

  it.skip('stdout is empty', async () => {
    const { stdout } = await factory().onclose
    expect(stdout).toBe('')
  })

  it.skip('stderr is empty', async () => {
    const { stderr } = await factory().onclose
    expect(stderr).toBe('')
  })
})
