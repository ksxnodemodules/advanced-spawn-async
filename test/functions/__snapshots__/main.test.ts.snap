// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when executable does not exist onclose promise 1`] = `
Object {
  "info": Object {
    "args": Array [],
    "command": "SomethingThatDoesNotExist",
    "error": [Error: spawn SomethingThatDoesNotExist ENOENT],
    "options": Object {},
    "process": Anything,
  },
  "name": "InternalError",
}
`;

exports[`when executable does not exist onexit promise 1`] = `
Object {
  "info": Object {
    "args": Array [],
    "command": "SomethingThatDoesNotExist",
    "error": [Error: spawn SomethingThatDoesNotExist ENOENT],
    "options": Object {},
    "process": Anything,
  },
  "name": "InternalError",
}
`;

exports[`when process successfully terminated with full arguments onclose promise 1`] = `
Object {
  "args": Array [
    "hello",
    "world",
  ],
  "command": "echo",
  "options": Object {
    "env": Object {
      "HELLO": "WORLD",
    },
  },
  "output": "hello world
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "hello world
",
}
`;

exports[`when process successfully terminated with full arguments onexit promise 1`] = `
Object {
  "args": Array [
    "hello",
    "world",
  ],
  "command": "echo",
  "options": Object {
    "env": Object {
      "HELLO": "WORLD",
    },
  },
  "output": "hello world
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "hello world
",
}
`;

exports[`when process successfully terminated with minimal arguments close promise matches snapshot 1`] = `
Object {
  "args": Array [],
  "command": "echo",
  "options": Object {},
  "output": "
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "
",
}
`;

exports[`when process successfully terminated with minimal arguments exit promise matches snapshot 1`] = `
Object {
  "args": Array [],
  "command": "echo",
  "options": Object {},
  "output": "
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "
",
}
`;

exports[`when process successfully terminated with minimal arguments process 1`] = `
Object {
  "onclose": Promise {},
  "onexit": Promise {},
  "process": Anything,
}
`;

exports[`when process successfully terminated with same event onclose promise 1`] = `
Object {
  "args": Array [
    "hello",
    "world",
  ],
  "command": "echo",
  "options": Object {
    "event": "close",
  },
  "output": "hello world
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "hello world
",
}
`;

exports[`when process successfully terminated with same event onexit promise 1`] = `
Object {
  "args": Array [
    "hello",
    "world",
  ],
  "command": "echo",
  "options": Object {
    "event": "close",
  },
  "output": "hello world
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "hello world
",
}
`;

exports[`when process successfully terminated with specified stdin close promise 1`] = `
Object {
  "args": Array [],
  "command": "bash",
  "options": Object {},
  "output": "stdin foo
stdin bar
stderr foo
stderr bar
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "stderr foo
stderr bar
",
  "stdout": "stdin foo
stdin bar
",
}
`;

exports[`when process successfully terminated with specified stdin exit promise 1`] = `
Object {
  "args": Array [],
  "command": "bash",
  "options": Object {},
  "output": "stdin foo
stdin bar
stderr foo
stderr bar
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "stderr foo
stderr bar
",
  "stdout": "stdin foo
stdin bar
",
}
`;

exports[`when process successfully terminated with stdout and stderr onclose promise 1`] = `
Object {
  "args": Array [
    Any<String>,
  ],
  "command": StringContaining "node",
  "options": Object {
    "env": Object {
      "HELLO": "WORLD",
    },
  },
  "output": "stdout[0]
stderr[0]
stdout[1]
stdout[2]
stderr[1]
stderr[2]
{ env: { HELLO: 'WORLD' } }
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "stdout[0]
stderr[0]
stdout[1]
stdout[2]
stderr[1]
stderr[2]
{ env: { HELLO: 'WORLD' } }
",
}
`;

exports[`when process successfully terminated with stdout and stderr onexit promise 1`] = `
Object {
  "args": Array [
    Any<String>,
  ],
  "command": StringContaining "node",
  "options": Object {
    "env": Object {
      "HELLO": "WORLD",
    },
  },
  "output": "stdout[0]
stderr[0]
stdout[1]
stdout[2]
stderr[1]
stderr[2]
{ env: { HELLO: 'WORLD' } }
",
  "process": Anything,
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "stdout[0]
stderr[0]
stdout[1]
stdout[2]
stderr[1]
stderr[2]
{ env: { HELLO: 'WORLD' } }
",
}
`;

exports[`when process terminated with non-zero status code onclose promise 1`] = `
Object {
  "info": Object {
    "args": Array [
      Any<String>,
    ],
    "command": StringContaining "node",
    "options": Object {
      "env": Object {
        "HELLO": "WORLD",
      },
    },
    "output": "stdout[0]
stderr[0]
stdout[1]
stdout[2]
stderr[1]
stderr[2]
{ env: { HELLO: 'WORLD' } }
",
    "process": Anything,
    "signal": null,
    "status": 123,
    "stderr": "",
    "stdout": "stdout[0]
stderr[0]
stdout[1]
stdout[2]
stderr[1]
stderr[2]
{ env: { HELLO: 'WORLD' } }
",
  },
  "name": "TerminationError",
}
`;

exports[`when process terminated with non-zero status code onexit promise 1`] = `
Object {
  "info": Object {
    "args": Array [
      Any<String>,
    ],
    "command": StringContaining "node",
    "options": Object {
      "env": Object {
        "HELLO": "WORLD",
      },
    },
    "output": "stdout[0]
stderr[0]
stdout[1]
stdout[2]
stderr[1]
stderr[2]
{ env: { HELLO: 'WORLD' } }
",
    "process": Anything,
    "signal": null,
    "status": 123,
    "stderr": "",
    "stdout": "stdout[0]
stderr[0]
stdout[1]
stdout[2]
stderr[1]
stderr[2]
{ env: { HELLO: 'WORLD' } }
",
  },
  "name": "TerminationError",
}
`;
